src/util/submit-addon.js

```diff
@@ -308,15 +313,80 @@ export default class Client {
     return promisify(pipeline)(contents, createWriteStream(destPath));
   }
 
-  async postNewAddon(
+  /*
+  This function aims to quickly hash the contents of the zip file that's being uploaded,
+  to compare it to the previous zip file that was uploaded, so we can skip the upload for
+  efficiency.
+
+  CRCs are used from the zip to avoid having to extract and hash  all the files.
+
+  Two zips that have different byte contents in their files must have a different hash;
+  but returning a different hash when the contents are the same in some cases is acceptable
+  - a false mismatch does not result in lost data.
+  */
+  async hashXpiCrcs(
+    filePath: string,
+    asyncFsReadFile: typeof defaultAsyncFsReadFile = defaultAsyncFsReadFile
+  ): Promise<string> {
+    const zip = await JSZip.loadAsync(
+      asyncFsReadFile(filePath, { createFolders: true })
+    );
+    const hash = createHash('sha256');
+    const entries = [];
+    zip.forEach((relativePath, entry) => {
+      let path = relativePath;
```

Comment:
Not sure if / is included or stripped, but if unsure, you can do this:

```suggestion
      let path = relativePath.replace(/\/+$/, "");
```

The source code suggests that this is needed for consistency: https://github.com/Stuk/jszip/blob/2ceb998e29d4171b4f3f2ecab1a2195c696543c0/lib/zipEntry.js#L143-L157